x/pairing/keeper/unresponsive_provider.go

```diff
@@ -111,84 +109,118 @@ func (k Keeper) getUnresponsiveProvidersToPunish(ctx sdk.Context, currentEpoch u
+func (k Keeper) updateCuForProviderInProviderCuCounterForUnreponsivenessMap(ctx sdk.Context, epoch uint64, epochsNumToCheckCUForUnresponsiveProvider uint64, epochsNumToCheckCUForComplainers uint64, providerStakeEntry epochstoragetypes.StakeEntry, providerCuCounterForUnreponsivenessMap map[string]providerCuCounterForUnreponsiveness) error {
 	epochTemp := epoch
+	createNewproviderCuCounterForUnreponsivenessEntryFlag := false
+	// get the provider's SDK account address
+	sdkStakeEntryProviderAddress, err := sdk.AccAddressFromBech32(providerStakeEntry.GetAddress())
+	if err != nil {
+		return utils.LavaFormatError("unable to sdk.AccAddressFromBech32(provider)", err, &map[string]string{"provider_address": providerStakeEntry.Address})
+	}
 
-	for counter := uint64(0); counter < epochsNumToGoBack; counter++ {
-		// Get previous epoch (from epochTemp)
-		previousEpoch, err := k.epochStorageKeeper.GetPreviousEpochStartForBlock(ctx, epochTemp)
-		if err != nil {
-			return 0, utils.LavaError(ctx, k.Logger(ctx), "get_previous_epoch", map[string]string{"err": err.Error(), "epoch": fmt.Sprintf("%+v", epochTemp)}, "couldn't get previous epoch")
-		}
+	// get the provider's providerCuCounterForUnreponsiveness
+	providerCuCounterForUnreponsivenessEntry, found, providerCuCounterForUnreponsivenessKey := k.getProviderCuCounterForUnreponsivenessEntryFromMap(providerStakeEntry, providerCuCounterForUnreponsivenessMap)
+	if !found {
+		// if providerCuCounterForUnreponsivenessEntry is not found, raise a flag to create new entry in providerCuCounterForUnreponsivenessMap
+		createNewproviderCuCounterForUnreponsivenessEntryFlag = true
+	}
 
-		// Get the current stake storages (from all chains)
-		stakeStorageList := k.getCurrentProviderStakeStorageList(ctx)
-		if len(stakeStorageList) == 0 {
-			// no provider is staked -> CU = 0
-			return 0, nil
-		}
+	// count the CU serviced by the unersponsive provider and used CU of the complainers
+	for counter := uint64(0); counter < epochsNumToCheckCUForUnresponsiveProvider+epochsNumToCheckCUForComplainers; counter++ {
+		// get providerPaymentStorageKey for epochTemp (other traits from the stake entry)
+		providerPaymentStorageKey := k.GetProviderPaymentStorageKey(ctx, providerStakeEntry.GetChain(), epochTemp, sdkStakeEntryProviderAddress)
 
-		// Get a providerPaymentStorage of all staked providers
-		providerPaymentStorageList, err := k.getProviderPaymentStorageFromStakeStorageList(ctx, previousEpoch, stakeStorageList)
-		if err != nil {
-			return 0, utils.LavaError(ctx, k.Logger(ctx), "get_provider_payment_storage_from_stake_storage_list", map[string]string{"err": err.Error(), "epoch": fmt.Sprintf("%+v", previousEpoch)}, "couldn't get providerPaymentStorageList from stakeStorageList")
+		// try getting providerPaymentStorage using the providerPaymentStorageKey
+		providerPaymentStorage, found := k.GetProviderPaymentStorage(ctx, providerPaymentStorageKey)
+		if !found {
```

Comment:
instead of this duplicate code wrap the whole count with the " if find" instead of the " if !find" and continue